src/wallet/wallet_tx_builder.cpp

```diff
@@ -0,0 +1,706 @@
+                assert(sendTo.has_value());
+                resolvedPayments.AddPayment(
+                        ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                changeAddr = sendTo.value();
+            },
+            [&](const libzcash::SproutPaymentAddress& addr) {
+                // for Sprout, we return change to the originating address using the tx builder.
+                changeAddr = addr;
+            },
+            [&](const libzcash::SproutViewingKey& vk) {
+                // for Sprout, we return change to the originating address using the tx builder.
+                changeAddr = vk.address();
+            },
+            [&](const libzcash::SaplingPaymentAddress& addr) {
+                // for Sapling, if using a legacy address, return change to the
+                // originating address; otherwise return it to the Sapling internal
+                // address corresponding to the UFVK.
+                std::optional<RecipientAddress> sendTo = sendFromAccount == ZCASH_LEGACY_ACCOUNT
+                    ? addr
+                    : pwalletMain->GenerateChangeAddressForAccount(
+                            sendFromAccount,
+                            allowedChangeTypes({ReceiverType::Sapling}));
+                assert(sendTo.has_value());
+                resolvedPayments.AddPayment(
+                        ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                changeAddr = sendTo.value();
+            },
+            [&](const libzcash::SaplingExtendedFullViewingKey& fvk) {
+                // for Sapling, if using a legacy address, return change to the
+                // originating address; otherwise return it to the Sapling internal
+                // address corresponding to the UFVK.
+                std::optional<RecipientAddress> sendTo = sendFromAccount == ZCASH_LEGACY_ACCOUNT
+                    ? fvk.DefaultAddress()
+                    : pwalletMain->GenerateChangeAddressForAccount(
+                            sendFromAccount,
+                            allowedChangeTypes({ReceiverType::Sapling}));
+                assert(sendTo.has_value());
+                resolvedPayments.AddPayment(
+                        ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                changeAddr = sendTo.value();
+            },
+            [&](const libzcash::UnifiedAddress& ua) {
+                auto zufvk = pwalletMain->GetUFVKForAddress(ua);
+                if (zufvk.has_value()) {
+                    auto sendTo = zufvk.value().GetChangeAddress(
+                            allowedChangeTypes(ua.GetKnownReceiverTypes()));
+                    assert(sendTo.has_value());
+                    resolvedPayments.AddPayment(
+                            ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                    changeAddr = sendTo.value();
```

Comment:
If we can get rid of the sole remaining `if` above, then all of these should be able to change to:
```suggestion
                    return sendTo.value();
```

Reply:
Did it!